### ISSUE TITLE
MODX 3.1.0 setup crashes if modUserGroupRole authority levels are not unique

### ISSUE THREAD

**Issue by BobRay** [modxcms/revolution#16669]
### Summary

I modUserGroupRole authority levels are not unique, setup crashes with this message:

"Multiple modUserGroup records with the same authority were found"

### Step to reproduce
Create modUserGroup records with the same authority value
Run setup.

I assume this is the result of this change:

"Make authority a unique index on modUserGroupRole (#16587)"

First of all, I think the message is incorrect since it refers to modUserGroup records rather than modUserGroupRole records. The modUserGroup table has no `authority` field.

More important, I can't think of any reason for requiring unique authority levels. 

I often use the same authority level for multiple user group roles. When users need different permissions, I put them in different user groups rather then giving them different authority levels. IMO, doing so makes it much easier to diagnose permission problems. I can't think of any harm that would be caused by having duplicate authority levels.

It also seems unreasonable to make current users edit their existing authority values in order to upgrade MODX, or to force them to make sure their authority levels are all unique. When you try to create a new role that uses an existing authority level, you get this message: `role_err_save`, and there's nothing in the description of the authority field that says the value has to be unique.

**Comment by opengeek** [comment]
The authority of a modUserGroupRole MUST be unique because that is the only reference back to a modUserGroupRole from any ACL record. This SHOULD have always been true, but seems to have been an oversight. There is no reason I can think of to have multiple modUserGroupRoles with the same authority. A "member" with authority of 9999 can be reused on every modUserGroup if you do not need different authority levels within a single group. This was a conscious decision, and the easiest way to resolve a problem in managing ACLs when authority values that were not unique were used to create the ACLs.

**Comment by BobRay** [comment]
Fair enough. It did catch me by surprise on several sites.  Still, the message needs to be corrected, possibly earlier in the process,  showing the correct object name.  And, I'd suggest giving a little more information about how to fix things so the use can complete the upgrade.

**Comment by BobRay** [comment]
There are still a couple of issues around this. First, if you manually try to save two roles with the same authority number, or save a role with an authority number that's already taken, it crashes the Manager. If you do it in code, it causes a 500 hundred error for the site. This definitely shouldn't happen.

Second, supposing I need to save a role, or multiple roles in an extra (I do). It seems nuts to pick an authority level then have an infinite loop that keeps checking to see if it exists, and keeps incrementing it by 1 until if finds one that isn't used, but I can't think of any other strategy.

**Comment by opengeek** [comment]
> There are still a couple of issues around this. First, if you manually try to save two roles with the same authority number, or save a role with an authority number that's already taken, it crashes the Manager. If you do it in code, it causes a 500 hundred error for the site. This definitely shouldn't happen.

Can you please enter an issue for the error occurring when you attempt to save a role with an existing authority level? It will get lost in this tangentally related issue.
 
> Second, supposing I need to save a role, or multiple roles in an extra (I do). It seems nuts to pick an authority level then have an infinite loop that keeps checking to see if it exists, and keeps incrementing it by 1 until if finds one that isn't used, but I can't think of any other strategy.

I cannot think of a single reason why an extra would need to involve roles at all. Roles were created for one reason, and one reason only. To allow a groups to define roles within them. This is a fairly finite thing. You are either a member or an admin in most cases I can think of. Could there be others? Sure. But those would apply across any user group. And in a vast majority of cases, there is no need for these roles at all. You are either a member of a group or you are not and that should suffice unless you have some serious enterprise level access control requirements.

**Comment by BobRay** [comment]
GoRevo copies all WebUsers and ManagerUsers members to the corresponding User Groups in Revo. I need a roles to assign them to those groups. I was hesitant to assign an authority level of 0 to all the Manager Users, since in Evo, the users roles determine what they can do in the Manager, so giving them all the same capabilities seemed wrong to me.  I did give them all an authority level of 0, in the latest version of GoRevo (not yet released), but created a limited policy for them in the new group. As you say, the WebUsers are no problem -- all getting 9999..

All that said, someone who uses authority levels could easily attempt to create multiple roles with the same authority level in the Manager (especially those coming from Evolution). They will get an error message, but the site now throws a 500 error, which shouldn't happen. It will also happen to GoRevo users who haven't yet upgraded to the latest version. 

I think I have other extras that may have this problem, but haven't had time to check.

BTW, I believe creating a role in code and leaving out the authority level gives the role an authority level of 0, which also seems wrong to me.

**Comment by opengeek** [comment]
I'm saying the roles you need already exist. I can't think of a single reason you need any role other than Member in the GoRevo package. And that role exists in the core already.

The bugs when trying to create a role in the manager obviously need to be fixed and that's why I asked for a separate issue for it.

**Comment by BobRay** [comment]
Closing. Will open a separate issue.
